cpp/ql/src/semmle/code/cpp/models/interfaces/FormattingFunction.qll

```diff
@@ -109,9 +109,24 @@ abstract class FormattingFunction extends ArrayFunction, TaintFunction {
   }
 
   /**
-   * Gets the position at which the output parameter, if any, occurs.
+   * Gets the position at which the output parameter, if any, occurs. This may
+   * be a buffer that characters are written to if this function behaves like
+   * `sprintf`.  Alternatively it may be a stream that is used for output if
+   * this function behaves like `fprintf` (see `isOutputStream`).
    */
   int getOutputParameterIndex() { none() }
+  
+  /**
+   * Holds if this function outputs to a global stream such as standard output,
+   * standard error or a system log. For example `printf`.
+   */
+  predicate isOutputGlobal() { none() }
+  
+  /**
+   * Holds if this function outputs to the stream indicated by
+   * `getOutputParameterIndex()`, that is, this function behaves like `fprintf`.
+   */
+  predicate isOutputStream() { none() }
```

Comment:
Could we just implement this as a `final` predicate that checks whether `getOutputParameterIndex()` has a result?

Reply:
No, this predicate only holds when `getOutputParameterIndex()` points to a stream parameter, not when it points to a buffer parameter.  I'm going to think about a better interface for this, or failing that at least better QLDoc...

Reply:
I've just pushed a commit combining `getOutputParameterIndex` and `isOutputStream` into one `getOutputParameterIndex(boolean isStream)`. I believe this makes it easier to understand what's going on and harder to miss the existence of `isOutputStream` by accident.

Thank you for the question that led me down this road.

Reply:
I like this interface a lot better. Do we need to keep a deprecated parameter-less version of `getOutputParameterIndex` around for compatibility?

Reply:
> Do we need to keep a deprecated parameter-less version of getOutputParameterIndex around for compatibility?

[There is one](https://github.com/github/codeql/pull/4754/files#diff-c81f331ad8237d9b6c277111ba00e160051d9c4ebdb36549cf27314bbe9b99f5R124), it doesn't really stand out in the diff.

Reply:
Ah, sorry. You're right!